### ISSUE TITLE
Sprite atlas bleeding with blade renderer

### ISSUE THREAD

**Issue by jansol** [zed-industries/zed#12352]
### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

![image](https://github.com/zed-industries/zed/assets/2588851/725b0078-ce19-4d68-8624-8f2336f29dbc)

Sporadically appears and disappears. May appear in one window but not another within the same Zed instance. Similar to #643 

### Environment

Zed: v1.0.0 (Zed Dev e19339bc1d038b39c5ee69ca728042d13d555fc7)
OS: Linux 1.0.0
Memory: 30.9 GiB
Architecture: x86_64

**Comment by apricotbucket28** [comment]
I think this may be related to https://github.com/zed-industries/zed/issues/11056 (probably the same cause), though the issues described are a bit different.

This is one of the more extreme cases I got while using Zed:
![image](https://github.com/zed-industries/zed/assets/71973804/135fc5a0-10b9-4d60-be50-a12812b64413)

It's also visible with the LSP loading indicator:
![image](https://github.com/zed-industries/zed/assets/71973804/402efd27-58f9-4224-a9dc-0276229f4b55)

**Comment by sk337** [comment]
i am having this issue as well

**Comment by apricotbucket28** [comment]
Hey @kvark, sorry to bother you! I'm not experienced with shaders, but this specific issue bothers me quite a lot. Do you have any guesses as to what could be causing this? e.g. could this be an issue with a specific primitive shader?

FWIW this may be related to https://github.com/zed-industries/zed/issues/12599

**Comment by kvark** [comment]
I'm not able to see it on my platform. It would help to get a GPU capture with RenderDoc and share it somewhere.

**Comment by apricotbucket28** [comment]
Thanks for the reply. I've used RenderDoc to generate two different captures (one at 100% scaling and another at 125%, where the issue is clearer).
[bleeding.zip](https://github.com/user-attachments/files/16487981/bleeding.zip)

**Comment by kvark** [comment]
That's very helpful! Unfortunately, RenderDoc doesn't properly associate resources with the draw calls for this case, so I'm not able to debug it in the capture:
![renderdoc-fail-resources](https://github.com/user-attachments/assets/9fa67fb0-eed4-458c-9230-c938ce3c1a1b)

Based on the captured data though, we are drawing right on the edge of pixel centers. This isn't expected. This logic is exactly matching between the original Metal backend and Blade, though. So my leading hypothesis is that the issue is in gpui itself and is platform-agnostic. The reason why we only see it on Linux is probably due to how the default texture contents (for the atlas) are not initialized to zero and happen to have opaque garbage regions:
![atlas](https://github.com/user-attachments/assets/714ef199-8573-43f6-bc12-3d2f186009a0)

I'll keep looking at this, but I expect the solution to be outside of Blade-specific code.

**Comment by kvark** [comment]
Awktually, a better idea - it's caused by incomplete handling (in gpui) of the fractional scaling. MacOS users/devices tend to not use/enable it as much, I suppose, hence the issue wasn't detected earlier.

**Comment by jansol** [comment]
But it also happens when display scale is 100%? My initial guess was that vulkan and metal sample textures differently (0,0 at the corner of the pixel vs in the center of it) but I haven't had time to dig into whether that is the case.

**Comment by kvark** [comment]
No the pixel center logic is exactly the same between then, this shouldn't be a difference here. Do you have a RenderDoc capture with 100% scaling by any chance? My expectation is that all coordinates that the vertex shaders operate on would be exactly on the pixel boundaries in this case.

For example, in "bleeding_125scaling" capture that @apricotbucket28 provided, I see Y vertex output absolute coordinates being 789.499035 to 771.998565. Clearly, the left boundary is problematic here.

The exact data is coming from `MonochromeSprite::bounds`. One suspect place is [Window::paint_glyph](https://github.com/zed-industries/zed/blob/49e736d8ef7db2701df0a15c996a3a395bde23d8/crates/gpui/src/window.rs#L2469) (or at least one of the places where the bounds are constructed), but it doesn't quite look like it could produce fractional coordinates.

**Comment by kvark** [comment]
I think I see what's happening in there. The layout is done in the coordinate system that doesn't match the screen. All coordinates get multiplied by `scale_factor` before going to the screen:
```rust
let scale_factor = self.scale_factor();
let bounds = bounds.scale(scale_factor);
```
In the content coordinates, we have primitives that are placed on the edge of the pixels. After multiplication, we get coordinates that end with .0, .25, .5, and .75.

Before these go into the `gl_Position` output, they also get divided by the screen size, and then the fixed hardware multiplies them back by the same size before rasterization. So this divide+multiply by a number (say, 897 in one of the captures) is pushing the ".5" resulting value slightly away from ".5". It could be ".4999034" for example, or ".500032". The difference is important though since it would determine if the pixel center is included in rasterization. We could address this by adding a small delta to the pixel coordinates if we know ".5" is our subpixel offset.

But that's only half of the problem. Even if ".5" was rasterized consistently, we'd still have another problem: if the pixel center happens to be too close to the primitive boundary, then the interpolated texture coordinate at this pixel will be too close to the atlas region boundary. Therefore, sampling this texel linearly will include the neighbors.
There are at least 2 ways to address this:
  1. Include a black transparent border (of 1 texel) around every atlas region. It costs in texture space and the allocator complexity.
  2. clamp the texture coordinates in the fragment shader, so that they are within half-texel from the boundary. That's what [we did in WebRender](https://github.com/mozilla/gecko-dev/blob/8b0666aff1197e1dd8017de366343de9c21ee437/gfx/wr/webrender/res/sample_color0.glsl#L23-L26). It costs in a bit of FS performance.

**Comment by kvark** [comment]
More concretely, here is a solution I'm thinking:
  1. provide all coordinates into the shader in the content space (even `u16` if possible, so that it's small).
  2. provide `scale_factor` separately in the global uniforms
  3. let the vertex shader figure out how to map this primitive onto the pixel grid, and also adjust the interpolated values like texture coordinates for consistency
  4. let the fragment shader clamp the interpolated texture coordinates to the inner area of the atlas region

**Comment by apricotbucket28** [comment]
Thanks for looking into this! My knowledge with graphics is little to none, and what you're describing seems like a much better solution.
Doing that might actually solve a lot of graphical issues as well, e.g. https://github.com/zed-industries/zed/issues/7436 (like @jansol mentioned) or other cases of bleeding that are still left (like the following)
![image](https://github.com/user-attachments/assets/834fa328-e703-4cc9-b690-d8d7c0efe340)

**Comment by JosephTLyons** [comment]
This is now included in today's [`v0.148.0-pre`](https://zed.dev/releases/preview/0.148.0) release, and will go out to stable next week.
